src/mesh/voxel_cloud.rs

```diff
@@ -228,6 +239,163 @@ impl VoxelCloud {
         }
     }
 
+    /// Computes boolean intersection (logical AND operation) of the current and
+    /// another Voxel cloud. The Voxel clouds have to intersect and the voxel
+    /// dimension must be equal. The current Voxel cloud will be mutated and
+    /// resized to the size and position of an intersection of the two Voxel
+    /// clouds (not the contained volumes).
+    pub fn boolean_intersection_mut(&mut self, other: &VoxelCloud) {
+        assert!(
+            approx::relative_eq!(self.voxel_dimensions, other.voxel_dimensions()),
+            "The two voxel clouds don't contain voxels of the same size"
+        );
+
+        // Find volume common to both voxel clouds.
+        // FIXME: This could be delegated to BoundingBox. @Optimization Compute
```

Comment:
1. I'm not sure we need both versions. In fact, there should be no difference between the two resulting voxel clouds. The cast there is just some block of space in a voxel cloud is only due to optimization reasons. The voxel cloud should be treated as endless, I think. Therefore I think we always want to shrink it to fit, unless the processing is more expensive than storing larger data. Also, what do you mean by _ compute the final bounding box from the volumes_ ? do you want to scan each source voxel cloud for volumes prior to the boolean operation? I think it is more expensive than calling `shrink_to_fit` at the end, isn't it?

Reply:
2. I think I don't understand

Reply:
3. we already had this - the constructor ordered inputs so that it was guaranteed that one point was min and the other was max. why is start + diagonal better when you need to check the diagonal to be all positive? it's easier to input two points and genrate proper min/max points. there can't be an error, whereas i'm not sure how would i treat a negative diagonal. 
i prefer returning to sorted points + we should implement more methods, like boolean operations (union, intersecion) of two bounding boxes and materialization of all 8 corners.

Reply:
4. exactly, i've been thinking the same. unless we want to convert all the cooridnates into cartesian, which should also work.

Reply:
5. okay

Reply:
i agree to remove the new separate

Reply:
> The voxel cloud should be treated as endless, I think. Therefore I think we always want to shrink it to fit, unless the processing is more expensive than storing larger data.

Interesting thought. In this case, yes, we would want only one version and they would be equivalent. We will need to implement custom `Eq` and document this.

> Also, what do you mean by _ compute the final bounding box from the volumes_ ? do you want to scan each source voxel cloud for volumes prior to the boolean operation? I think it is more expensive than calling shrink_to_fit at the end, isn't it?

Yes. You are possibly right from the time perspective, but resizes fragment memory and the scan would still be linear. We can leave it for a `FIXME: @Optimization` though.

Reply:
> 2. I think I don't understand

I imagine we could have `BoundingBox::intersect(other: &BoundingBox) -> BoundingBox`, and similar for other boolean operations. Of course, it would behave a bit weird, so maybe we don't want to use the boolean operator nomenclature, but it seems clear enough to me.

The result would always be just one bounding box.

Reply:
> why is start + diagonal better when you need to check the diagonal to be all positive? it's easier to input two points and genrate proper min/max points. there can't be an error, whereas i'm not sure how would i treat a negative diagonal.

You are right that one is not worse than the other - they are exactly equivalent. Either way you have to _correctly generate_ the min/max points or origin+diagonal.

> i prefer returning to sorted points

Just to clarify, by sorted you mean constructed from the individual components, right? Because sorting is not sufficient.

> + we should implement more methods, like boolean operations (union, intersecion) of two bounding boxes and materialization of all 8 corners.

This is my point number 2.

Reply:
> exactly, i've been thinking the same. unless we want to convert all the cooridnates into cartesian, which should also work.

We could theoretically go either way, but having voxel-space (integer) bounding boxes will not force you to convert between the two coord spaces even more. I already think we have some subtle errors in the conversions we will have to hunt down later, so I'd rather not add more.